crates/wasi-provider/src/lib.rs

```diff
@@ -83,6 +90,127 @@ impl<S: Store + Send + Sync> WasiProvider<S> {
+                    // been validated by the k8s API
+                    Ok((vol.deref().clone(), Some(guest_path)))
+                })
+                .collect::<anyhow::Result<HashMap<PathBuf, Option<PathBuf>>>>()
+        } else {
+            Ok(HashMap::default())
+        }
+    }
+
+    async fn run_one_container(
+        &self,
+        container: &KubeContainer,
+        pod: &Pod,
+        client: &kube::client::Client,
+        modules: &mut HashMap<String, Vec<u8>>,
+        volumes: &HashMap<String, Ref>,
+        container_handles: &mut HashMap<
+            String,
+            kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>,
+        >,
+    ) -> anyhow::Result<()> {
+        let env = Self::env_vars(&container, pod, &client).await;
+        let args = container.args.clone().unwrap_or_default();
+        let module_data = modules
+            .remove(&container.name)
+            .expect("FATAL ERROR: module map not properly populated");
+        let container_volumes = Self::volume_path_map(container, volumes)?;
+
+        let runtime = WasiRuntime::new(
+            module_data,
+            env,
+            args,
+            container_volumes,
+            self.log_path.clone(),
+        )
+        .await?;
+
+        debug!("Starting container {} on thread", container.name);
+        let handle = runtime.start().await?;
+        container_handles.insert(container.name.clone(), handle);
+
+        Ok(())
+    }
+
+    async fn run_one_container_to_completion(
+        &self,
+        container: &KubeContainer,
+        pod: &Pod,
+        client: &kube::client::Client,
+        modules: &mut HashMap<String, Vec<u8>>,
+        volumes: &HashMap<String, Ref>,
+        container_handles: &mut HashMap<
+            String,
+            kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>,
+        >,
+    ) -> anyhow::Result<()> {
+        self.run_one_container(container, pod, client, modules, volumes, container_handles)
+            .await?;
+        let result = self
+            .wait_for_termination(container, container_handles)
+            .await?;
+        // TODO: should we remove the container?
+        if result.succeeded {
+            Ok(())
+        } else {
+            Err(anyhow::anyhow!(
+                "Init container {} failed with message {}",
+                container.name,
+                result.message
+            ))
+        }
+    }
+
+    async fn wait_for_termination(
+        &self,
+        container: &KubeContainer,
+        container_handles: &mut HashMap<
+            String,
+            kubelet::container::Handle<wasi_runtime::Runtime, wasi_runtime::HandleFactory>,
+        >,
+    ) -> anyhow::Result<ContainerTerminationResult> {
+        let handle = container_handles.get(&container.name).unwrap();
+        let mut status = handle.status();
+        loop {
+            // TODO: clippy wants us to do it this way but ugh
```

Comment:
What is the other way you are trying to do it here?

Reply:
I think it was a `match` but can't remember now.  I just don't like the way the destructuring of the Terminated enum breaks up the flow.  I'll take another look.

_copy of "Refactoring" eerily falls open at the entry for 'Extract Method'_

Reply:
Here we go, during refactoring I made it the way I wanted it to look and:

```rust
warning: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`
   --> crates/wasi-provider/src/lib.rs:193:13
    |
193 | /             match status {
194 | |                 Some(kubelet::container::Status::Terminated {
195 | |                     message, failed, ..
196 | |                 }) => {
...   |
202 | |                 _ => (),
203 | |             }
    | |_____________^
```

With the refactoring and the explaining variable, though, the `if let` won't look quite as horrid...

Reply:
Update: it still looks foul #damnyouclippy